src/sql/src/pure.rs

```diff
@@ -1292,6 +1297,341 @@ async fn purify_alter_source(
     })
 }
 
+async fn purify_create_table_from_source(
+    catalog: impl SessionCatalog,
+    mut stmt: CreateTableFromSourceStatement<Aug>,
+    storage_configuration: &StorageConfiguration,
+) -> Result<PurifiedStatement, PlanError> {
+    let scx = StatementContext::new(None, &catalog);
+    let CreateTableFromSourceStatement {
+        name: _,
+        columns,
+        constraints,
+        source: source_name,
+        if_not_exists: _,
+        external_reference,
+        with_options,
+    } = &mut stmt;
+
+    // Columns and constraints cannot be specified by the user but will be populated below.
+    if !columns.is_empty() {
+        sql_bail!("CREATE TABLE .. FROM SOURCE column definitions cannot be specified directly");
+    }
+    if !constraints.is_empty() {
+        sql_bail!(
+            "CREATE TABLE .. FROM SOURCE constraint definitions cannot be specified directly"
+        );
+    }
+
+    // Get the source item
+    let item = match scx.get_item_by_resolved_name(source_name) {
+        Ok(item) => item,
+        Err(e) => return Err(e),
+    };
+
+    // Ensure it's an ingestion-based and alterable source.
+    let desc = match item.source_desc()? {
+        Some(desc) => desc.clone().into_inline_connection(scx.catalog),
+        None => {
+            sql_bail!("cannot ALTER this type of source")
+        }
+    };
+    let unresolved_source_name: UnresolvedItemName = source_name.full_item_name().clone().into();
+    let qualified_source_name = item.name();
+    let connection_name = desc.connection.name();
+
+    let crate::plan::statement::ddl::TableFromSourceOptionExtracted {
+        text_columns,
+        ignore_columns,
+        details,
+        seen: _,
+    } = with_options.clone().try_into()?;
+    assert_none!(details, "details cannot be explicitly set");
+
+    // Our text column values are unqualified (just column names), but the purification methods below
+    // expect to match the fully-qualified names against the full set of tables in upstream, so we
+    // need to qualify them using the external reference first.
+    let qualified_text_columns = text_columns
+        .iter()
+        .map(|col| {
+            UnresolvedItemName(
+                external_reference
+                    .0
+                    .iter()
+                    .chain_one(col)
+                    .map(|i| i.clone())
+                    .collect(),
```

Comment:
This is again making me nervous. How do we guarantee that this has the right number of levels? Can we ever end up with a 4 level `UnresolvedItemName` here if the external reference is fully qualified? What happens if we do?

Reply:
This is allowed to be less-than fully qualified or fully-qualified, since each source implementation normalizes these values after resolving them using the appropriate `SourceReferenceResolver` (this is the remaining use-case for that struct). In postgres the 4-level `UnresolvedItemName` is fine since the column-name is always popped off before the rest of the qualifier is resolved against the publication tables:
https://github.com/MaterializeInc/materialize/blob/f880946bb5a89ffcc1f086430e4172bea15e139e/src/sql/src/pure/postgres.rs#L69-L90

I think this is safer than some of the previous stuff where we've had to manipulate `UnresolvedItemName`s since this is all pre-purification and will error out if resolution fails before storing anything in the catalog.

Also once we can rip out subsources we can simplify this logic to make the methods accept the column-names directly rather than having to pre-qualify them using the external reference.